File: examples/05-fork-join/Example11/Example11.java
// =================================================================
//
// File: Example11.java
// Author(s):
//				 A01651517 Peter Glez
//				 A01703947 Juan Alcántara
// Description: This file implements the code that transforms a
//				grayscale image using Java's Fork-Join.
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

import java.awt.image.BufferedImage;
import java.io.File;
import javax.imageio.ImageIO;
import java.io.IOException;
import java.util.concurrent.RecursiveAction;
import java.util.concurrent.ForkJoinPool;

public class Example11 extends RecursiveAction {
    private static final int MIN = 15_000;
	private int src[], dest[], start, end, width, height;

    public Example11(int src[], int dest[], int width, int height, int start, int end) {
        this.src = src;
        this.dest = dest;
        this.start = start;
        this.end = end;
    }

    private void grayScale(int index) {
        int red, green, blue, avg;
        red = (src[index]&0x00FF0000)>>16;
        green = (src[index]&0x0000FF00)>>8;
        blue = (src[index]&0x000000FF);
        avg = (red + green + blue)/3;
        dest[index] = 0xFF000000 | (avg << 16) | (avg << 8) | avg;
    }

    protected void computeDirectly() {
        int index, ren, col;
        for (index = start; index < end; index++)
            grayScale(index);
    }

    @Override
    protected void compute() {
        if ((start - end) <= MIN) {
            computeDirectly();
        } else {
            int mid = start + ((end - start) / 2);
            invokeAll(new Example11(src, dest, width, height, start, mid),
                new Example11(src, dest, width, height, mid, end));
        }
    }

    public static void main(String args[]) throws Exception {
        long startTime, stopTime;
        double ms;
        int src[], dest[], w, h;
        ForkJoinPool pool;

        if (args.length != 1) {
            System.out.println("usage: java Example11 image_file");
            System.exit(-1);
        }

        final String fileName = args[0];
        File srcFile = new File(fileName);
        final BufferedImage source = ImageIO.read(srcFile);

        w = source.getWidth();
        h = source.getHeight();
        src = source.getRGB(0, 0, w, h, null, 0, w);
        dest = new int[src.length];

        System.out.printf("Starting with %d threads...\n", Utils.MAXTHREADS);
        ms = 0;
        for (int i = 0; i < Utils.N; i++) {
            startTime = System.currentTimeMillis();
            pool = new ForkJoinPool(Utils.MAXTHREADS);
            pool.invoke(new Example11(src, dest, w, h, 0, (w*h)));

            stopTime = System.currentTimeMillis();
            ms += (stopTime - startTime);
        }

        System.out.printf("avg time = %.5f\n", (ms / Utils.N));
		final BufferedImage destination = new BufferedImage(w, h, BufferedImage.TYPE_INT_ARGB);
		destination.setRGB(0, 0, w, h, dest, 0, w);

		javax.swing.SwingUtilities.invokeLater(new Runnable() {
            public void run() {
               ImageFrame.showImage("Original - " + fileName, source);
            }
        });

		javax.swing.SwingUtilities.invokeLater(new Runnable() {
            public void run() {
               ImageFrame.showImage("Blur - " + fileName, destination);
            }
        });

		try {
			ImageIO.write(destination, "png", new File("gray.jpg"));
			System.out.println("Image was written succesfully.");
		} catch (IOException ioe) {
			System.out.println("Exception occured :" + ioe.getMessage());
		}
    }
}

File: examples/06-openmp/example11.cpp
// =================================================================
//
// File: example11.cpp
// Author(s):
//        A01651517 Pedro González
//				A01703947 Juan Alcántara
// Description: This file implements the code that transforms a
//				grayscale image. Using OpenCV and OpenMP, to compile:
//				g++ example11.cpp `pkg-config --cflags --libs opencv4` -fopenmp
//
// Copyright (c) 2020 by Tecnologico de Monterrey.
// All Rights Reserved. May be reproduced for any non-commercial
// purpose.
//
// =================================================================

#include <stdio.h>
#include <stdlib.h>
#include <opencv2/opencv.hpp>
#include <opencv2/core/core.hpp>
#include <opencv2/highgui/highgui.hpp>
#include <opencv2/highgui/highgui_c.h>
#include <opencv2/imgproc/imgproc.hpp>
#include <opencv2/imgcodecs/imgcodecs.hpp>
#include "utils.h"

void gray_scale(cv::Mat *src, cv::Mat *dest) {
  int i;
  #pragma omp parallel shared(src, dest, i)
  for (i = 0 ; i < src->rows*src->cols ; i++) {
    float calc = (
        (float) src->at<cv::Vec3b>(i)[RED]+
        (float) src->at<cv::Vec3b>(i)[GREEN]+
        (float) src->at<cv::Vec3b>(i)[BLUE]
    )/3;

    dest->at<cv::Vec3b>(i)[RED] = calc;
    dest->at<cv::Vec3b>(i)[GREEN] = calc;
    dest->at<cv::Vec3b>(i)[BLUE] = calc;
  }
}

int main(int argc, char* argv[]) {
	int i;
	double acum;

	if (argc != 2) {
	printf("usage: %s source_file\n", argv[0]);
		return -1;
	}

	cv::Mat src = cv::imread(argv[1], cv::IMREAD_COLOR);
	cv::Mat dest = cv::Mat(src.rows, src.cols, CV_8UC3);
	if (!src.data) {
	printf("Could not load image file: %s\n", argv[1]);
		return -1;
	}

	acum = 0;
	for (i = 0; i < N; i++) {
		start_timer();

		// call the implemented function
    gray_scale(&src, &dest);

		acum += stop_timer();
	}

	printf("avg time = %.5lf ms\n", (acum / N));

	cv::namedWindow("Original", cv::WINDOW_AUTOSIZE);
    cv::imshow("Original", src);

	cv::namedWindow("Gray", cv::WINDOW_AUTOSIZE);
    cv::imshow("Gray", dest);

	cv::waitKey(0);
	cv::imwrite("gray_scale.png", dest);

	return 0;
}
